### PR TITLE
Add Pipeline Parallel (and 2D PP+FSDP) support

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #318

runs PP+DP and PP+TP without issue,
runs PP+TP+DP with decreasing loss, but fails DCP save

Supports only simple schedules currently, gpipe and 1f1b.

Ads cmdline/toml arg for specifiying split points, in a unified
way between tracer or manual frontend.

  e.g. user can specifiy "layers.2,layers.4" as split points.

Currently uses manual frontend by default, but allows specifying
tracer frontend.  Tracer frontend requires working around additional
compatibility limitations, indicated by raising assertions, and is
not ready for wider use  yet.

### CODE REVIEW

File: torchtitan/config_manager.py [hunk 1]
```diff
@@ -207,6 +207,62 @@ def __init__(self):
...
+
+                For simple schedules, this should be 1.
+
+                For looped schedules, this can be greater than one.
+
+                If the number of stages produced by splitting does not match the expected number of stages,
+                an error will be raised for sanity.""",
+        )
+        self.parser.add_argument(
+            "--experimental.pipeline_parallel_split_points",
```
Comment (tianyu-l) [inline torchtitan/config_manager.py:226]: Am I right that in some vague sense: out of the three configs `training.pipeline_parallel_degree`, `experimental.pipeline_parallel_stages_per_rank`, and `experimental.pipeline_parallel_split_points`, given any two, the remaining one can be inferred?

As a user I may not care too much about where the exact split points are (as long as it's roughly even), so can we just determine the split points given pp-degree and number of stages per rank.

File: torchtitan/config_manager.py [hunk 2]
```diff
@@ -207,6 +207,62 @@ def __init__(self):
...
+            default=[],
+            help="""
+                Specify names of modules to use as the beginning of a split point.
+
+                e.g. ["layers.1"] will cause the model to be split into 2 stages,
+                the first containing all the layers up to layers.0,
+                the second containing layers.1 and all the remaining layers.
+
+                Note: fully-automated splitting may be enabled in the future,
+                but currently the split points must be specified manually for both manual and tracer.""",
```
Comment (tianyu-l) [inline torchtitan/config_manager.py:239]: oh, just saw this. If that's the case, either we can remove the `pipeline_parallel_stages_per_rank` option and print some thing to user, or keep it as sanity check. Maybe better to keep.

File: torchtitan/config_manager.py (hunk 1 above)
Comment (wconstab) [reply torchtitan/config_manager.py:226]: I think to keep things simple, since we did not yet support virtual stages or looped schedules in torchtitan, this argument (num_stages) can be deleted for now.

In the future, I think we don't even need to add this num_stages arg unless we support automated splitting.  If the user gives split points (as is required now), we can require the number of split points be divisible by the number of ranks for looped schedules or equal the number of ranks for regular schedules.

File: torchtitan/parallelisms/parallelize_llama.py [hunk 3]
```diff
@@ -129,15 +145,281 @@ def get_tp_parallel_strategy(
...
+        fqns, job_config.experimental.pipeline_parallel_split_points, pp_rank
+    )
+    assert len(new_names) == len(this_stage_layer_names), (
+        len(new_names),
+        len(this_stage_layer_names),
+    )
+    for n, n_ in zip(new_names, this_stage_layer_names):
+        assert n == n_, (n, n_)
+
+    input_seqlen = 2048  # TODO hack
```
Comment (wconstab) [inline torchtitan/parallelisms/parallelize_llama.py]: blocker: need to fix this. perhaps by moving freqs_cis handling in the real model code?

File: torchtitan/parallelisms/parallelize_llama.py [hunk 4]
```diff
@@ -129,15 +145,281 @@ def get_tp_parallel_strategy(
...
+    # layers of the model that map to this stage, not the whole model.
+
+    # Get example input
+    if pp_rank == 0:
+        input_shape = (job_config.training.batch_size, job_config.training.seq_len)
+        input = torch.randint(
+            model_config.vocab_size, input_shape, dtype=torch.int64, device=device
+        )
+
+        # HACK- can't use shape inference via execution of the PP stage inside ManualPipelineStage API, becuase the
```
Comment (wconstab) [inline torchtitan/parallelisms/parallelize_llama.py]: proposal to land this part as-is. We need to add the lazy shape inference part to manual stage api, but i think we can clean this up after it lands.

File: train.py [hunk 5]
```diff
@@ -144,6 +147,15 @@ def main(job_config: JobConfig):
         dp_rank = dp_mesh.get_local_rank()
     else:
         dp_degree, dp_rank = 1, 0
+
+    if parallel_dims.pp_enabled:
+        pp_mesh = world_mesh["pp"]
+        pp_degree = pp_mesh.size()
+        pp_rank = pp_mesh.get_local_rank()
```
Comment (tianyu-l) [inline train.py]: defined but not used?

File: train.py [hunk 6]
```diff
@@ -144,6 +147,15 @@ def main(job_config: JobConfig):
...
+
+    else:
```
Comment (tianyu-l) [inline train.py]: can we remove this `else` branch?

File: torchtitan/parallelisms/parallelize_llama.py [hunk 7]
```diff
@@ -222,13 +389,17 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
                 )
             # As an optimization, do not reshard after forward for the last
             # transformer block since FSDP would prefetch it immediately
-            reshard_after_forward = int(layer_id) < len(model.layers) - 1
+            # reshard_after_forward = int(layer_id) < len(model.layers) - 1
+            # TODO(whc) need to fix correctly handle layer-ids on pp-split module
+            reshard_after_forward = True
```
Comment (wconstab) [inline torchtitan/parallelisms/parallelize_llama.py]: @awgu I'm confused, i think you mentioned i need to set this to true a while back.  In most of the PP schedules, i probably want to disable resharding so i get zero-2, don't I?

Why would i want this to be set to true?

Comment (awgu) [reply torchtitan/parallelisms/parallelize_llama.py]: I might have misspoke. We want to set this to `False` for PP.

File: train.py [hunk 8]
```diff
@@ -216,10 +241,13 @@ def loss_fn(pred, labels):
         f"({gpu_mem_stats.max_reserved_pct:.2f}%)"
     )
 
+    if isinstance(model, FSDPModule) and parallel_dims.pp_enabled:
+        # reshard now to counteract an issue where FSDP's states got advanced during PP stage shape inference
+        model.reshard()
```
Comment (wconstab) [inline train.py]: @awgu following up, is this actually a 'bug' in FSDP or is this something we need to sort out at the trainer level (or inside DCP)? Should I file an issue to track?

Comment (awgu) [reply train.py]: I think this is not a bug in FSDP per se but a bug in the composition of the PP shape inference and FSDP.

If PP shape inference runs a fake forward, then _any_ module that changes global state in the forward would have a similar issue.

Let me try to understand this better though. (sorry for the lack of understanding of the PP implementation)
- Is the PP stage shape inference orthogonal to the frontend tracer vs. manual splitting?
- It looks like we get `stage, model` --> apply TP/AC/FSDP --> build pipeline schedule. Is the shape inference happening in the "build pipeline schedule" part? (This seems to be the only way it makes sense that FSDP's states are getting advanced.)

Comment (wconstab) [reply train.py]: for manual stage (which is what i'm currently iterating on) we do shape inference during stage creation (__init__) today, but i'm proposing we move it to be lazy on first forward.  That's primarily so that we dont have to specify input shapes to each stage, for UX reasons, and so we also don't have to specify an input shape at all during stage creation (which is awkward to do any time before the first batch comes out of the dataloader).

for tracer, we need to specify the input shape for the model for the tracer to run, and that will always happen at stage creation time rather than later.

after applying FSDP, we should not actually be doing any fake-forward stuff today.

File: torchtitan/parallelisms/parallelize_llama.py [hunk 9]
```diff
@@ -129,15 +134,187 @@ def get_tp_parallel_strategy(
...
+    layers_per_rank = len(model.layers) // parallel_dims.pp
+    split_spec = {
+        f"layers.{i * layers_per_rank}": SplitPoint.BEGINNING
+        for i in range(1, parallel_dims.pp)
+    }
+
+    # Create a pipeline representation from the model
+    pipe = pipeline(
+        model,
+        parallel_dims.pp,
```
Comment (kwen2501) [inline torchtitan/parallelisms/parallelize_llama.py]: nit: can you add a #TODO here to say that it should be a configurable # of microbatches? thx

File: torchtitan/parallelisms/pipelining_utils.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+    else:
+        raise NotImplementedError(
+            f"{job_config.experimental.pipeline_parallel_schedule} is not implemented"
+        )
+    logger.info(
+        f"Using pipeline schedule {job_config.experimental.pipeline_parallel_schedule}"
+    )
+    return schedule_class(
+        stage,
+        n_microbatches=parallel_dims.pp,
```
Comment (kwen2501) [inline torchtitan/parallelisms/pipelining_utils.py]: nit: same #TODO

**Review by kwen2501 — APPROVED** [review]
LGTM.
Is there a conclusion with `model.reshard()`?

File: torchtitan/parallelisms/parallelize_llama.py [hunk 11]
```diff
@@ -129,15 +134,187 @@ def get_tp_parallel_strategy(
...
+        )
+    else:
+        # later layers (assume all start w/ a transformer layer)
+        input = torch.rand(
+            size=(
+                job_config.training.batch_size,
+                int(job_config.training.seq_len // parallel_dims.tp),
+                model_config.dim,
+            ),
+            dtype=torch.bfloat16 if parallel_dims.dp_enabled else torch.float32,
```
Comment (kwen2501) [inline torchtitan/parallelisms/parallelize_llama.py]: I think `dtype=torch.bfloat16 if parallel_dims.dp_enabled` is a bit arbitrary.

File: torchtitan/parallelisms/parallelize_llama.py [hunk 12]
```diff
@@ -129,15 +134,187 @@ def get_tp_parallel_strategy(
...
+        )
+    else:
+        # earlier layers (assume all end in a transformer layer)
+        output = torch.rand(
+            size=(
+                job_config.training.batch_size,
+                int(job_config.training.seq_len // parallel_dims.tp),
+                model_config.dim,
+            ),
+            dtype=torch.bfloat16 if parallel_dims.dp_enabled else torch.float32,
```
Comment (kwen2501) [inline torchtitan/parallelisms/parallelize_llama.py]: Same here.

File: train.py (hunk 8 above)
Comment (kwen2501) [reply train.py]: In the **current** state of code, since you are feeding known input shapes to `ManualPipelineStage` constructor, why is there a need for shape inference? Or, did `ManualPipelineStage` just run it anyway? It seems to me that the easiest way to land the **current** state of code is make that shape inference conditional and remove `model.reshard()` here. (Considering that developing lazy shape inference may take a while.)

File: torchtitan/parallelisms/parallelize_llama.py (hunk 11 above)
Comment (wconstab) [reply torchtitan/parallelisms/parallelize_llama.py]: if your point is that its fragile to do shape inference before we have the full/real model to do inference on, i agree-  but if we put inference into the runtime, I think that would be solved.

or, are you saying there might be a cleaner way to decide this (like adding a config option for runtime dtype, and then checking that instead of just checking 'dp enabled'?)  maybe that's a good idea.

File: torchtitan/parallelisms/parallelize_llama.py (hunk 9 above)
Comment (wconstab) [reply torchtitan/parallelisms/parallelize_llama.py]: oh yea. should we make it a cmd line arg?  i think that would be good, but it should have a default value imo.  and i think the default value should be related to the number of stages.

File: train.py (hunk 8 above)
Comment (wconstab) [reply train.py]: i need to double check what the reshard is for.

maybe previously, i was running shape-inference _and_ the manual stage creation was after fsdp wrap.

but now, i am doing stage creation before fsdp wrap, and I can’t use its shape inference there anyway because it would infer the wrong shapes at that early point.

Comment (wconstab) [reply train.py]: 2D with DP ran fine locally when I removed this. its removed on the PR now, make sure CI is still happy..

File: torchtitan/parallelisms/parallelize_llama.py (hunk 9 above)
Comment (kwen2501) [reply torchtitan/parallelisms/parallelize_llama.py]: For titan, I don't know if a good default value is one that _just works_ or one that _gives good perf_.
For the former you can choose PP degree. For the latter you may want to choose one bigger than PP degree. It also depends on whether the schedule is interleaved or not. Interleaved schedules may require less microbatches.

File: torchtitan/parallelisms/parallelize_llama.py (hunk 11 above)
Comment (kwen2501) [reply torchtitan/parallelisms/parallelize_llama.py]: Sorry for not being clear enough. Yeah, I was more referring to the latter, i.e. `dp_enabled` does not necessarily mean `bfloat16` -- just in case an `==` assumption leads to unexpected effects.

Comment (wconstab) [reply torchtitan/parallelisms/parallelize_llama.py]: Yea I think we need to expose a dtype config and use that for both places.

File: torchtitan/config_manager.py [hunk 13]
```diff
@@ -210,10 +214,68 @@ def __init__(self):
...
+                and finally wrapped in a PipelineStage.  tracer frontend is currently more experimental.""",
+        )
+        self.parser.add_argument(
+            "--experimental.pipeline_parallel_microbatches",
+            type=int,
+            default=None,
+            help="""
+                How many microbatches to split the full training batch into when using pipeline parallelism.
+
+                The overall training batch size must be evenly divisible by the number of microbatches.
```
Comment (wanchaol) [inline torchtitan/config_manager.py]: "The overall training batch size" could we be more specific? either use global batch size or local batch size?

File: torchtitan/config_manager.py [hunk 14]
```diff
@@ -437,6 +499,10 @@ def parse_args_from_command_line(
                 aux_parser.add_argument(
                     "--" + arg, action="store_true" if val else "store_false"
                 )
+            elif arg == "experimental.pipeline_parallel_split_points":
+                # type inference breaks here, since the type is just 'list' and it ends up flattening
```
Comment (wanchaol) [inline torchtitan/config_manager.py]: do you mean without this code change, it would break?

File: torchtitan/parallelisms/parallelize_llama.py [hunk 15]
```diff
@@ -129,15 +134,192 @@ def get_tp_parallel_strategy(
...
+            size=(
+                job_config.training.batch_size,
+                int(job_config.training.seq_len // parallel_dims.tp),
+                model_config.dim,
+            ),
+            dtype=TORCH_DTYPE_MAP[job_config.training.mixed_precision_param]
+            if parallel_dims.dp_enabled
+            else torch.float32,
+            device=device,
+        )
```
Comment (wanchaol) [inline torchtitan/parallelisms/parallelize_llama.py]: Can we encapsulate L209-L258 into a `input_output_shape_inference` function and put that into pipeline_utils?

File: test_runner.py [hunk 16]
```diff
@@ -100,22 +174,42 @@ def build_test_list(args):
...
+
+        if test_flavor.requires_seed_checkpoint:
+            dump_folder_arg = None
+            for arg in override_arg:
+                if "--job.dump_folder" in arg:
+                    dump_folder_arg = arg
+            assert (
+                dump_folder_arg is not None
+            ), "Can't use seed checkpoint if folder is not specified"
+            print("Creating seed checkpoint")
```
Comment (wanchaol) [inline test_runner.py]: let's use log instead of print?

File: torchtitan/parallelisms/parallelize_llama.py [hunk 17]
```diff
@@ -129,15 +134,192 @@ def get_tp_parallel_strategy(
     return RowwiseParallel, ColwiseParallel
 
 
+def _llama_fqns(num_layers):
```
Comment (wanchaol) [inline torchtitan/parallelisms/parallelize_llama.py]: wondering if this is really needed? i.e. we can just pass model to `split_stage_fqns` and then looping over `model.named_childrens` to get all the fqn?

File: torchtitan/parallelisms/parallelize_llama.py [hunk 18]
```diff
@@ -129,15 +134,192 @@ def get_tp_parallel_strategy(
...
+    return (tokens,)
+
+
+def pipeline_llama_manual(
+    model, world_mesh, parallel_dims, job_config: JobConfig, device, model_config: Dict
+):
+    """
+    This API gets individual torch.nn.Module objects for each pipeline stage (including virtual stages).
+
+    The SPMD parallelisms should be applied to
```
Comment (wanchaol) [inline torchtitan/parallelisms/parallelize_llama.py]: missing comment?

**Review by wanchaol — APPROVED** [review]
This looks reasonablely good to me, have some additional comments inlined.

File: torchtitan/parallelisms/parallelize_llama.py (hunk 15 above)
Comment (wconstab) [reply torchtitan/parallelisms/parallelize_llama.py]: i think those lines are llama-model specific? thats why i kept them here.  (The plan is to get rid of these still by providing a runtime shape inference thing, but it might not be the highest priority).

**Comment by wanchaol** [comment]
@wconstab looks like CI is failing now, is it because the APIs for PP not in nightly yet? If so we should probably wait until the nightly is there and then reland this
